### PR TITLE
Cls2 547 generate task overdue notifications

### PR DESCRIPTION
**Pull request by PaulWheatcroft**
### Description of change

Send a reminder when the due date for a ticket is one day in the past. This will run on a cron scheduler at 8 AM

### Testing
Pull this branch, run in docker end ensure there is a Task that is one day overdue.

Run python in the terminal then copy and paste the following...

>import os
from datetime import datetime
from logging import getLogger
import django
import environ
os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'config.settings.local')
'config.settings.local'
django.setup()
from datahub.task.tasks import generate_reminders_tasks_overdue
generate_reminders_tasks_overdue()

This will trigger the routine to send a reminder. You will be able to see in the the email routine trigger within the API and rq logs.

### Checklist

* [x] Has this branch been rebased on top of the current `main` branch?

  <details>
  <summary>Explanation</summary>
  
  The branch should not be stale or have conflicts at the time reviews are requested.
  
  </details>

* [ ] Is the CircleCI build passing?

### General points

<details>
<summary><strong>Other things to check</strong></summary><p></p>

* Make sure `fixtures/test_data.yaml` is maintained when updating models
* Consider the admin site when making changes to models
* Use select-/prefetch-related field lists in views and search apps, and update them when fields are added
* Make sure the README is updated e.g. when adding new environment variables

</details>

See [docs/CONTRIBUTING.md](https://github.com/uktrade/data-hub-api/blob/main/docs/CONTRIBUTING.md) for more guidelines.

### CODE REVIEW

File: datahub/task/tasks.py [hunk 1]
```diff
@@ -479,3 +494,103 @@ def send_task_email(adviser, task, reminder, update_task, email_template_class):
         update_task=update_task,
         reminders=[reminder],
     )
+
+
+# start here
```
Comment (chopkinsmade) [inline datahub/task/tasks.py:499]: Can remove this

File: datahub/task/tasks.py [hunk 2]
```diff
@@ -479,3 +494,103 @@ def send_task_email(adviser, task, reminder, update_task, email_template_class):
...
+        wait=False,
+    ) as acquired:
+        if not acquired:
+            logger.info(
+                'Reminders for tasks overdue are already being processed by another worker.',
+            )
+            return
+        now = timezone.now()
+        yesterday = datetime.date.today() - datetime.timedelta(days=1)
+        tasks = Task.objects.filter(due_date=yesterday)
```
Comment (chopkinsmade) [inline datahub/task/tasks.py:529]: You're getting every task with due date of yesterday, need to ignore any completed ones. and add some texts around this ignoring completed ones

File: datahub/task/tasks.py [hunk 3]
```diff
@@ -479,3 +494,103 @@ def send_task_email(adviser, task, reminder, update_task, email_template_class):
         update_task=update_task,
         reminders=[reminder],
     )
+
+
+# start here
+
+
```
Comment (samuele-mattiuzzo) [inline datahub/task/tasks.py:501]:
```suggestion
```

File: datahub/task/tasks.py (hunk 1 above)
Comment (PaulWheatcroft) [reply datahub/task/tasks.py:499]: Yep indeed I should have

File: datahub/task/test/test_tasks.py [hunk 4]
```diff
@@ -1060,3 +1104,241 @@ def test_schedule_advisers_added_to_task_adds_job_to_queue_for_each_adviser(
...
+        )
+
+        assert expected_message in caplog.messages
+
+        if lock_acquired:
+            mock_notify_adviser_by_rq_email.assert_called_once()
+        else:
+            mock_notify_adviser_by_rq_email.assert_not_called()
+
+    def test_generate_reminders_for_overdue_tasks(
```
Comment (chopkinsmade) [inline datahub/task/test/test_tasks.py:1156]: There's quite a lot to understand what this test is doing, to make it obvious that archived should be excluded can you add another test called `test_generate_reminders_for_overdue_tasks_excludes_archived`, that creates a single archived task and checks no reminders are sent

Comment (PaulWheatcroft) [reply datahub/task/test/test_tasks.py:1156]: Done.

**Review by chopkinsmade — APPROVED** [review]

**Review by samuele-mattiuzzo — APPROVED** [review]

**Pull request merged**